torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#if defined(__cplusplus) && (__cplusplus >= 201703L)
+#include <filesystem>
+namespace fs = std::filesystem;
+#elif defined(__cplusplus) && (__cplusplus >= 201402L)
+#include <experimental/filesystem>
+namespace fs = std::experimental::filesystem;
+#else
+#error "C++14 or Higher is required for filesystem library!"
+#endif
+
+#include <unordered_map>
+#include <vector>
+
+#include <c10/macros/Export.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//! KernelDbEntry captures information to be printed per fusion in a csv file
+//! that is used to restore a hash map
+struct KernelDbEntry {
+  //! Cuda kernel function signature that is required to load the Cubin
+  std::string kernel_name;
+  //! Compilation args supplied to NVRTC -- register usage and compute
+  //! capability can be specific to a kernel instance
+  std::string compile_args;
+  //! Full file path to Cuda Kernel
+  std::string kernel_code_file;
+  //! Full file path to cubin
+  std::string cubin_file;
+};
+
+//! KernelDb class is a singleton structure that is used to open, query, and
+//! write to the the database that is held in a hash map.  The kernel code is
+//! used as string key to the hash map.
+class TORCH_CUDA_CU_API KernelDb {
+  KernelDb(bool _disabled);
+
+  KernelDb(const KernelDb&) = delete;
+  KernelDb& operator=(const KernelDb&) = delete;
+
+  //! Open is private because this method should only be called once by the
+  //! singleton upon creation to create a new db or restore an existing one.
+  bool open(
+      const std::string& kernel_db_dir,
+      const std::string& kernel_db_file,
+      bool use_temp_dir);
+
+ public:
+  //! Thread-Safe method to get the Meyer's singleton -- Interface
```

Comment:
Thread safe meaning even multiple torch processes are supported on the same file system?